server/src/main/java/org/elasticsearch/index/codec/vectors/DefaultIVFVectorsWriter.java

```diff
@@ -0,0 +1,686 @@
+            return ords[idx];
+        }
+
+        float getScore(int idx) {
+            return scores[idx];
+        }
+
+        int size() {
+            return idx;
+        }
+    }
+
+    // TODO unify with OSQ format
+    static class BinarizedFloatVectorValues {
+        private OptimizedScalarQuantizer.QuantizationResult corrections;
+        private final byte[] binarized;
+        private final byte[] initQuantized;
+        private float[] centroid;
+        private final FloatVectorValues values;
+        private final OptimizedScalarQuantizer quantizer;
+
+        private int lastOrd = -1;
+
+        BinarizedFloatVectorValues(FloatVectorValues delegate, OptimizedScalarQuantizer quantizer) {
+            this.values = delegate;
+            this.quantizer = quantizer;
+            this.binarized = new byte[discretize(delegate.dimension(), 64) / 8];
+            this.initQuantized = new byte[delegate.dimension()];
+        }
+
+        public OptimizedScalarQuantizer.QuantizationResult getCorrectiveTerms(int ord) {
+            if (ord != lastOrd) {
+                throw new IllegalStateException(
+                    "attempt to retrieve corrective terms for different ord " + ord + " than the quantization was done for: " + lastOrd
+                );
+            }
+            return corrections;
+        }
+
+        public byte[] vectorValue(int ord) throws IOException {
+            if (ord != lastOrd) {
+                binarize(ord);
+                lastOrd = ord;
+            }
+            return binarized;
+        }
+
+        private void binarize(int ord) throws IOException {
+            corrections = quantizer.scalarQuantize(values.vectorValue(ord), initQuantized, INDEX_BITS, centroid);
+            packAsBinary(initQuantized, binarized);
+        }
+    }
+
+    static class OffHeapCentroidAssignmentScorer implements CentroidAssignmentScorer {
+        private final IndexInput centroidsInput;
+        private final int numCentroids;
+        private final int dimension;
+        private final float[] scratch;
+        private float[] q;
+        private final long centroidByteSize;
+        private int currOrd = -1;
+
+        OffHeapCentroidAssignmentScorer(IndexInput centroidsInput, int numCentroids, FieldInfo info) {
+            this.centroidsInput = centroidsInput;
+            this.numCentroids = numCentroids;
+            this.dimension = info.getVectorDimension();
+            this.scratch = new float[dimension];
+            this.centroidByteSize = dimension + 3 * Float.BYTES + Short.BYTES;
+        }
+
+        @Override
+        public int size() {
+            return numCentroids;
+        }
+
+        @Override
+        public float[] centroid(int centroidOrdinal) throws IOException {
+            if (centroidOrdinal == currOrd) {
+                return scratch;
+            }
+            centroidsInput.seek(numCentroids * centroidByteSize + (long) centroidOrdinal * dimension * Float.BYTES);
```

Comment:
nit: may be save `numCentroids * centroidByteSize` as variable rather then recalculating it every time